File: programmers/java/src/level2/kakao/kakao2022/ParkingFeeCal.java
package level2.kakao.kakao2022;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// https://programmers.co.kr/learn/courses/30/lessons/92341
public class ParkingFeeCal {
    public static void main(String[] args) {
        ParkingFeeCal pfc = new ParkingFeeCal();
        System.out.println(Arrays.toString(pfc.solution(new int[]{180, 5000, 10, 600}, new String[]{"05:34 5961 IN", "06:00 0000 IN", "06:34 0000 OUT", "07:59 5961 OUT", "07:59 0148 IN", "18:59 0000 IN", "19:09 0148 OUT", "22:59 5961 IN", "23:00 5961 OUT"})));
    }

    public Integer[] solution(int[] fees, String[] records) {
        int minTime = fees[0];
        int minFee = fees[1];
        int unitTime = fees[2];
        int unitFee = fees[3];

        int[] inTime = new int[10000];
        Arrays.fill(inTime, -2);
        int[] total = new int[10000];
        for (String r : records) {
            String[] r1 = r.split(" ");
            String[] r2 = r1[0].split(":");
            int h = Integer.parseInt(r2[0]);
            int m = Integer.parseInt(r2[1]);
            int t = h * 60 + m;
            if (r1[2].equals("IN")) {
                inTime[Integer.parseInt(r1[1])] = t;
            } else {
                int num = Integer.parseInt(r1[1]);
                total[Integer.parseInt(r1[1])] += t - inTime[num];
                inTime[num] = -1;
            }
        }

        List<Integer> answer = new ArrayList<>();
        for (int i = 0; i < 10000; i++) {
            if (inTime[i] != -2) {
                if (inTime[i] != -1) {
                    total[i] += 23 * 60 + 59 - inTime[i];
                }
                if (total[i] <= minTime) {
                    answer.add(minFee);
                } else {
                    answer.add(minFee + (int) (Math.ceil((total[i] - minTime) / (double)unitTime) * unitFee));
                }
            }
        }
        return answer.toArray(new Integer[0]);
    }
}

File: programmers/python/level2/kakao/2022/parking_fee_cal.py
# https://programmers.co.kr/learn/courses/30/lessons/92341?language=python3
import math

def solution(fees, records):
    min_time, min_fee, unit_time, unit_fee = fees
    info = {}
    answer = {}
    for record in records:
        print(record)
        time, num, status = record.split(" ")
        num = int(num)
        if status == "IN" :
            info[num] = time
        else:
            in_h, in_m = map(int, info[num].split(":"))
            out_h, out_m = map(int, time.split(":"))
            time = (out_h * 60 + out_m) - (in_h * 60 + in_m)
            if num in answer:
                answer[num] += time
            else:
                answer[num] = time
            del(info[num])
        print(answer)

    if len(info) > 0 :
        for i in info.keys():
            print(i, info[i])
            in_h, in_m = map(int, info[i].split(":"))
            time = (23 * 60 + 59) - (in_h * 60 + in_m)
            if i in answer:
                answer[i] += time
            else:
                answer[i] = time
            print(answer)


    total = []
    for ans in answer.keys():
        print(ans, answer[ans])
        time = answer[ans]
        if time <= min_time :
            total.append((ans, min_fee))
        else:
            total.append((ans, min_fee + math.ceil((time - min_time) / unit_time) * unit_fee))

    total.sort()
    answer = []
    for t in total:
        num, fee = t
        answer.append(fee)
    return answer

# print(solution([180, 5000, 10, 600], ["05:34 5961 IN", "06:00 0000 IN", "06:34 0000 OUT", "07:59 5961 OUT", "07:59 0148 IN", "18:59 0000 IN", "19:09 0148 OUT", "22:59 5961 IN", "23:00 5961 OUT"]))
print(solution([120, 0, 60, 591], ["16:00 3961 IN","16:00 0202 IN","18:00 3961 OUT","18:00 0202 OUT","23:58 3961 IN"]))
# print(solution([1, 461, 1, 10], ["00:00 1234 IN"]))
